### PR TITLE
sketch of how default kernel will change

### PR DESCRIPTION
**Pull request by R-M-Lee**
We started discussing a little bit here:
https://github.com/experimental-design/bofire/discussions/430

Briefly: we will change the default priors (actually it's a bit more than that... the default kernel really) to reflect the findings of the paper [Vanilla Bayesian Optimization Performs Great in High Dimensions](https://arxiv.org/html/2402.02229v3) and the corresponding changes being made by the botorch team ([Update the default SingleTaskGP prior](https://github.com/pytorch/botorch/commit/83bb30c13db10075d180e5f7083f1b9d5e54e188) and [Use Standardize by default for SingleTaskGP](https://github.com/pytorch/botorch/commit/bcdea09cf9ae6ac2ab4099641c602b932d0b945f))

Here is a first very rough commit to check... I think that this would have the desired effect:

- dimensionality scaled lengthscale priors
- new noise prior
- no output scaling but output is standardized by default

If this makes sense (@jduerholt ) then I can clear it up and make the associated changes e.g., in SingleTaskGPHyperconfig.

Proposal: we just change the values of `BOTORCH_NOISE_PRIOR` and `BOTORCH_LENGTHCALE_PRIOR`, use them as defaults, and remove the `HVARFNER` stuff

### CODE REVIEW

**Comment by jduerholt** [comment]
Hi Robert,

looks good for me.

Let us rename the botorch priors to `ThreeSixPrior` as they are often called like this in the literature.

Why not keep the `Hvarfner` name? It is from his paper ;)

The question would then which priors to include in the hyperopt? We could go for all three (MBO, ThreeSix, Hvarfner) and combine it with a an additional input called `outputscale` to toggle in the hyperopt if the kernel should be combined with an outputscale or not? But this could also be done in a seperate PR.

The current PR is failing due to the lengthscale feature importance calculation, as it is expecting a lengthscale based kernel inside a scale kernel. It has to be adjusted so that it also understand it, when it is a pure kernel. But this is just one furhter if expresion.

Best,

Johannes

File: bofire/data_models/surrogates/single_task_gp.py [hunk 1]
```diff
@@ -99,16 +101,13 @@ class SingleTaskGPSurrogate(TrainableBotorchSurrogate):
...
 
     kernel: AnyKernel = Field(
-        default_factory=lambda: ScaleKernel(
-            base_kernel=MaternKernel(
-                ard=True,
-                nu=2.5,
-                lengthscale_prior=BOTORCH_LENGTHCALE_PRIOR(),
-            ),
-            outputscale_prior=BOTORCH_SCALE_PRIOR(),
+        default_factory=lambda: MaternKernel(
```
Comment (jduerholt) [inline bofire/data_models/surrogates/single_task_gp.py:104]: In the original paper and I think also in botorch they are using an RBF Kernel, or?

**Comment by R-M-Lee** [comment]
Thanks for the feedback. Ok then I'll clean this up and ask you to have another look later

**Comment by R-M-Lee** [comment]
Hi @jduerholt (maybe also interesting for @bertiqwerty),

I have not yet modified the length scale feature importance calculation or the hyperopt config... I just wanted to convince myself that the new priors work well. So I used the new multinorm benchmark problem with 20 dimensions to see whether the new priors give a performance boost. In this one test, they actually make it worse (this is a maximization problem)
 
image

Ideas? Either a different problem to confirm the performance or something I am missing in the surrogate definition (although it seems simple enough to me). Maybe I should use one of the test functions from the Hvarfner paper

The script I used to generate the data in the picture is [here](https://github.com/experimental-design/bofire/pull/435/commits/96e920116027022a8105d1a5a59799c87da102fb)

**Comment by jduerholt** [comment]
Hi @R-M-Lee,

very interesting finding.

When I implemented the priors orignally (which could be of course buggy), I tested it only on ZDT1 and saw there some improvement in comparison to the usual priors. You can find it here: https://github.com/experimental-design/bofire/blob/main/tutorials/benchmarks/011-ZDT1.ipynb

SAASBO still outperforms it there. Another high-dimensional function that we have in BoFire is the thirty dimensional branin function (https://github.com/experimental-design/bofire/blob/main/tutorials/benchmarks/006-30dimBranin.ipynb). I am currently running some tests there. Unfortunately, they have other functions in the paper. So, I think it makes sense to run our implementation on one of the functions from the paper.

Other possibility would be to run plain botorch on the benchmark from above. What we definitely can check is that the botorch model created by our surrogate looks the same (kernel, priors etc.) as the one which is currently setup in default botorch (using the Hvarfner priors)

Best,

Johannes

**Comment by jduerholt** [comment]
For the 30dim Branin, the new priors perform slighlty better: 
image

**Comment by R-M-Lee** [comment]
hmm, I'm still not sure enough with the new priors to finalize this pr... here is the result of using the Hartmann function in 25D (only the first 6 dimensions are influential, others are just to inflate the dimensionality). I have plotted the distribution of responses at each iteration, _not_ the "best so far". So it's clear from this that the Hvarfner priors, at least how we have them implemented, lead to a lot more exploration. Maybe this is the desired behavior; there's an argument in the Hvarfner paper about EI + short lengthscale priors leading to proposals close to the incumbent - maybe that's just what we see in this plot

image

Here is the typical staircase (cummin) plot. Now the new and old priors are pretty indistinguishable:

image

I should probably use the same init random samples to compare the methods. Will try to remember that next time

Next I will compare bofire surrogate with botorch surrogate as you suggest.

**Comment by R-M-Lee** [comment]
Ok I did the comparison now with some random train and test data. I think the results are pretty convincing that we have it implemented correctly:

image

**Comment by jduerholt** [comment]
> so I'll clean up, take care of the feature importance functions and the hyperconfig, then you can review

Thank you very much!

**Comment by R-M-Lee** [comment]
> The question would then which priors to include in the hyperopt? We could go for all three (MBO, ThreeSix, Hvarfner) and combine it with a an additional input called outputscale to toggle in the hyperopt if the kernel should be combined with an outputscale or not? But this could also be done in a seperate PR.

I did not change the hyperopt part yet. But I think your suggestion is good and we should do it like that. As you said, we can easily do this in a separate PR

**Comment by R-M-Lee** [comment]
Hi @jduerholt, quick reminder: this is still waiting for your review

**Comment by jduerholt** [comment]
Oh, sorry. Totally forgot it. I will review it tomorrow. Can you make the branch up do date with the recent changes in BoFire?

**Comment by R-M-Lee** [comment]
ok, I think it's up-to-date now so feel free to give your thoughts when you can @jduerholt

**Review by jduerholt — APPROVED** [review]
Looks good to me. Thank you! Only some minor comments. Best, Johannes

File: bofire/benchmarks/single.py [hunk 2]
```diff
@@ -608,14 +664,24 @@ def __init__(
                     cov_mat[:, i] = 0.0
                     cov_mat[i, i] = 10.0
                 gaussians.append(multivariate_normal(mean=mean, cov=cov_mat))
-
+        for i in range(n_gaussians):
+            prefactors.append(
+                (2 * np.pi) ** (-dim / 2) / np.sqrt(np.linalg.det(gaussians[i].cov))
+            )
         self.gaussians = gaussians
+        self.prefactors = prefactors
```
[the remark this answers is not in the document]
Comment (R-M-Lee) [reply bofire/benchmarks/single.py:672]: It's not a change to Branin, although how github displays the diff makes it look that way. It's a minor addition to the `Multinormalpdfs` benchmark. That one is a mixture of Gaussians and I just defined a prefactor for each Gaussian in the mixture. I am being a bit naughty by smuggling it in here, but I thought it's a minor change to a benchmark that probably no-one is using yet

File: bofire/data_models/surrogates/mixed_single_task_gp.py [hunk 3]
```diff
@@ -60,9 +60,9 @@ def _update_hyperparameters(
             )
         else:
             noise_prior, lengthscale_prior, _ = (
-                BOTORCH_NOISE_PRIOR(),
-                BOTORCH_LENGTHCALE_PRIOR(),
-                BOTORCH_SCALE_PRIOR(),
+                THREESIX_NOISE_PRIOR(),
+                THREESIX_LENGTHSCALE_PRIOR(),
+                THREESIX_SCALE_PRIOR(),
```
[the remark this answers is not in the document]
Comment (R-M-Lee) [reply bofire/data_models/surrogates/mixed_single_task_gp.py:65]: I'll open a separate PR for this

**Pull request merged**